File: packages/modules/api/go/src/api/crossref.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

type Author struct {
	Given  string `json:"given"`
	Family string `json:"family"`
}

type DateParts struct {
	DateParts [][]int `json:"date-parts"`
}

type CrossRefMessage struct {
	Author         []Author  `json:"author"`
	Title          []string  `json:"title"`
	ContainerTitle []string  `json:"container-title"`
	Volume         string    `json:"volume"`
	Issue          string    `json:"issue"`
	Page           string    `json:"page"`
	PublishedPrint *DateParts `json:"published-print"`
	PublishedOnline *DateParts `json:"published-online"`
}

type CrossRefResponse struct {
	Message CrossRefMessage `json:"message"`
}

type Reference struct {
	ID      string
	Authors []Author
	Title   string
	Journal string
	Volume  string
	Issue   string
	Pages   string
	Year    int
	URL     string
}

type CrossRefClient struct {
	BaseURL string
}

func NewCrossRefClient() *CrossRefClient {
	return &CrossRefClient{BaseURL: "https://api.crossref.org"}
}

func (c *CrossRefClient) GetWork(id string) (*Reference, error) {
	client := &http.Client{Timeout: 60 * time.Second}
	url := fmt.Sprintf("%s/works/%s", c.BaseURL, id)
	
	resp, err := client.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("crossref.org API error: %s", resp.Status)
	}

	var data CrossRefResponse
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}

	m := data.Message
	published := m.PublishedPrint
	if published == nil {
		published = m.PublishedOnline
	}
	
	year := 0
	if published != nil && len(published.DateParts) > 0 && len(published.DateParts[0]) > 0 {
		year = published.DateParts[0][0]
	}

	title := ""
	if len(m.Title) > 0 {
		title = m.Title[0]
	}

	journal := ""
	if len(m.ContainerTitle) > 0 {
		journal = m.ContainerTitle[0]
	}

	return &Reference{
		ID:      id,
		Authors: m.Author,
		Title:   title,
		Journal: journal,
		Volume:  m.Volume,
		Issue:   m.Issue,
		Pages:   m.Page,
		Year:    year,
		URL:     fmt.Sprintf("https://doi.org/%s", id),
	}, nil
}

File: packages/modules/api/rust/src/api/crossref.rs
use serde::{Deserialize, Serialize};
use reqwest::Client;
use std::time::Duration;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Author {
    pub given: Option<String>,
    pub family: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DateParts {
    #[serde(rename = "date-parts")]
    pub date_parts: Vec<Vec<i32>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CrossRefMessage {
    pub author: Vec<Author>,
    pub title: Vec<String>,
    #[serde(rename = "container-title")]
    pub container_title: Vec<String>,
    pub volume: Option<String>,
    pub issue: Option<String>,
    pub page: Option<String>,
    #[serde(rename = "published-print")]
    pub published_print: Option<DateParts>,
    #[serde(rename = "published-online")]
    pub published_online: Option<DateParts>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CrossRefResponse {
    pub message: CrossRefMessage,
}

#[derive(Debug)]
pub struct Reference {
    pub id: String,
    pub authors: Vec<Author>,
    pub title: String,
    pub journal: String,
    pub volume: Option<String>,
    pub issue: Option<String>,
    pub pages: Option<String>,
    pub year: i32,
    pub url: String,
}

pub struct CrossRefClient {
    base_url: String,
    client: Client,
}

impl CrossRefClient {
    pub fn new() -> Self {
        Self {
            base_url: "https://api.crossref.org".to_string(),
            client: Client::builder()
                .timeout(Duration::from_secs(60))
                .build()
                .unwrap(),
        }
    }

    pub async fn get_work(&self, id: &str) -> Result<Reference, reqwest::Error> {
        let url = format!("{}/works/{}", self.base_url, id);
        let resp = self.client.get(url).send().await?.error_for_status()?;
        let data = resp.json::<CrossRefResponse>().await?;

        let m = data.message;
        let published = m.published_print.as_ref().or(m.published_online.as_ref());
        
        let mut year = 0;
        if let Some(p) = published {
            if let Some(first_part) = p.date_parts.get(0) {
                if let Some(y) = first_part.get(0) {
                    year = *y;
                }
            }
        }

        Ok(Reference {
            id: id.to_string(),
            authors: m.author,
            title: m.title.first().cloned().unwrap_or_default(),
            journal: m.container_title.first().cloned().unwrap_or_default(),
            volume: m.volume,
            issue: m.issue,
            pages: m.page,
            year,
            url: format!("https://doi.org/{}", id),
        })
    }
}
